docs/LgcOverview.md

```diff
@@ -0,0 +1,199 @@
+method, such as `CreateImageSample`, BuilderRecorder records the call in the IR at the insertion
+point as a call to an `llpc.call.*` method, in this example `llpc.call.image.sample`. Such a
+call is a varargs call, so BuilderRecorder only needs to overload the declarations by return
+type, not by arg types.
+
+Then, the first pass run in `Pipeline::Generate` is the BuilderReplayer pass. That pass
+creates its own BuilderImpl, and replays each recorded call into it, generating the same
+IR as would have been generated if the front-end had chosen to use BuilderImpl directly.
+
+### BuilderImpl
+
+If the front-end selects BuilderImpl mode for Builder, or when running the BuilderReplayer
+pass, the Builder is an instance of BuilderImpl. That generates IR more directly for
+Builder methods such as `CreateImageSample`. BuilderImpl multiple-inherits from a bunch of classes
+that implement the different categories of Builder calls (arithmetic, descriptors, image,
+shader input/output, matrix, misc, subgroup), and each of those is a subclass of BuilderImplBase,
+a subclass of Builder.
+
+### PipelineState
+
+The Pipeline object created by the front-end calling `BuilderContext::CreatePipeline` is in
+fact an instance of PipelineState, an LGC-internal subclass of Pipeline. PipelineState contains
+the outside-IR state used through the middle-end.
+
+BuilderImpl code can access PipelineState directly, because BuilderImpl has a pointer to it.
+
+Middle-end passes (patch passes) access PipelineState via the PipelineStateWrapper pass.
+PipelineStateWrapper is an immutable pass, which is like a module pass but is assumed by LLVM's
+pass manager to never be invalidated.
+
+PipelineState stores two kinds of outside-IR state:
+
+* The state set by the front-end calling methods in Pipeline, such as the vertex input descriptions.
+  For a pipeline compile using BuilderRecorder, this state gets written into the IR as metadata
+  in `PipelineState::Link`, and read out the first time PipelineState is used in the middle-end.
+  Thus this kind of state can be considered a cache of information that is in IR metadata, and
+  in theory could be re-read if necessary at any point in the pass flow, although that does not
+  actually happen. Where this would happen is if a command-line compiler writes out the IR
+  after completing the front-end, and then a separate command invocation runs the middle-end.
+
+* The second kind of outside-IR state is the state set in middle-end passes (and BuilderImpl).
+  This is `ResourceUsage` and `InterfaceData`, which are declared in their own separate file
+  `llpcResourceUsage.h`.
+  This state does not get written into IR metadata. In theory we could add a facility to write
+  the state into IR metadata and read it back, and that would allow us to support lit testing of
+  individual middle-end passes with `-stop-before` and `-run-pass` options like `llc`. But
+  that does not happen now.
+
+### ShaderModes
```

Comment:
I don't understand what ShaderModes represent? Which info will it record?

Reply:
Added more info there.